src/next_array.rs

```diff
@@ -0,0 +1,80 @@
+use core::mem::MaybeUninit;
+
+/// Helper struct to build up an array element by element.
+struct ArrayBuilder<T, const N: usize> {
+    arr: [MaybeUninit<T>; N],
+    i: usize
+}
+
+impl<T, const N: usize> ArrayBuilder<T, N> {
+    pub fn new() -> Self {
+        Self { arr: maybe_uninit::uninit_array(), i: 0 }
+    }
+    
+    pub unsafe fn push_unchecked(&mut self, x: T) {
+        debug_assert!(self.i < N);
+        *self.arr.get_unchecked_mut(self.i) = MaybeUninit::new(x);
+        self.i += 1;
+    }
+    
+    pub fn take(mut self) -> Option<[T; N]> {
+        if self.i == N {
+            unsafe {
+                // SAFETY: prevent double drop.
+                self.i = 0;
+                // SAFETY: [MaybeUninit<T>; N] and [T; N] have the same layout.
+                let init_arr_ptr = &self.arr as *const _ as *const [T; N];
+                Some(core::ptr::read(init_arr_ptr))
```

Comment:
This needs a SAFETY comment citing what the preconditions of `ptr::read` are, and proving why they are satisfied.